### ISSUE TITLE
port names should match endpoint names

### ISSUE THREAD

**Issue by kadel** [redhat-developer/odo#4737]
Assuming that devfile containers have following endpoints
```
      endpoints:
        - name: first
          targetPort: 3000
        - name: second
          targetPort: 4000
```

The k8s Service that gets generated for odo should have matching port names

Current situation:
```
  ports:
  - name: port-3000
    port: 3000
    protocol: TCP
    targetPort: 3000
  - name: port-4000
    port: 4000
    protocol: TCP
    targetPort: 4000
```

/priority high
/kind bug

**Comment by mik-dass** [comment]
The service port naming is done here https://github.com/devfile/library/blob/5d88bd3c1220097161ef07d78c9fe7605be24ab7/pkg/devfile/generator/utils.go#L229.

**Comment by kadel** [comment]
> The service port naming is done here https://github.com/devfile/library/blob/5d88bd3c1220097161ef07d78c9fe7605be24ab7/pkg/devfile/generator/utils.go#L229.

We should coordinate with the Devfiles team to change this.

**Comment by mik-dass** [comment]
Currently we allow multiple URLs on the same port, but k8s doesn't allow services to have two ports with the same port number and protocol combination.

**Comment by kadel** [comment]
This problem was again raised by user on slack

> I'm using odo 2.3.1 with OpenShift 4.9 and trying to set the Service port name in the devfile, but it wont stick. I always get a default value of name: "port-8080" in my Service when what I want is a port name of "http" (for Service  Mesh and Kiali usage). Here is my devfile snippet, any ideas what I need to set:
> components:
> - container:
>     endpoints:
>     - exposure: public
>       name: "http"
>       path: /
>       targetPort: 8080

**Comment by valaparthvi** [comment]
>The service port naming is done here https://github.com/devfile/library/blob/5d88bd3c1220097161ef07d78c9fe7605be24ab7/pkg/devfile/generator/utils.go#L229.

These changes were made to accommodate another issue that odo was facing, see #4060, https://github.com/devfile/api/issues/287, and https://github.com/devfile/library/pull/51.

**Comment by yangcao77** [comment]
> Currently we allow multiple URLs on the same port, but k8s doesn't allow services to have two ports with the same port number and protocol combination.

This is also because another issue that tools will face. As devfile allows multiple endpoints within same component to use same port. but you can only expose a port number once in a K8s service. In other word, if it behavior as described in the main issue description, the K8s service spec generated will be invalid :

```
      endpoints:
        - name: first
          targetPort: 3000
        - name: second
          targetPort: 3000
```

The k8s Service that gets generated is valid:

```
  ports:
  - name: first
    port: 3000
    protocol: TCP
    targetPort: 3000
  - name: second
    port: 3000
    protocol: TCP
    targetPort: 3000
```

**Comment by yangcao77** [comment]
> > The service port naming is done here https://github.com/devfile/library/blob/5d88bd3c1220097161ef07d78c9fe7605be24ab7/pkg/devfile/generator/utils.go#L229.
> 
> These changes were made to accommodate another issue that odo was facing, see #4060, [devfile/api#287](https://github.com/devfile/api/issues/287), and [devfile/library#51](https://github.com/devfile/library/pull/51).

This is not the same issue with [devfile/api#287](https://github.com/devfile/api/issues/287), and [devfile/library#51](https://github.com/devfile/library/pull/51). those were pod container port name
The service port naming conversion was done in this PR: https://github.com/devfile/library/commit/ccd9d189c1885bf3e2ee69ab8cb57eb69917b9af#diff-9126bba4cbc803f4a34bfee5d830ed5af0ed5ebb1450bd7f2ebd4f554ea15a97
And IIRC, we were trying to resolve the issue mention in my last comment: https://github.com/redhat-developer/odo/issues/4737#issuecomment-983777050

**Comment by yangcao77** [comment]
To summarize the current issue: 
The devfile Spec & Odo:
- Endpoint has limitation of 63 chars length
- allows multiple endpoints under the same component shares the same port number

K8s:
- pod container port name has to be within 15 chars length
- service cannot have duplicate port exposure
This was the reasons we decided to use our own naming conversion for pod container port name and service port name.


Now, a Odo user reports the service port name does not match the actual endpoint name.  And we are looking for a more appropriate solution for the container port name & service port name.

**Comment by kadel** [comment]
> * allows multiple endpoints under the same component shares the same port number

Can we make this invalid? I'm not sure if there is a case where it would be required to have endpoints exposing the same port on one container component

**Comment by elsony** [comment]
Need to check to make sure it doesn't causes problems on update/migration

**Comment by valaparthvi** [comment]
* Remove the duplicate endpoint support from odo
* Check if limiting the endpoint char length from 63 to 15 has any side effect on update/migration.

**Comment by valaparthvi** [comment]
Once https://github.com/devfile/api/pull/702 is merged, we should add validation check to `odo url create` to ensure that the name length does not exceed 15 before writing the information to devfile otherwise it will cause an error every time the devfile is parsed.

Scope of the issue:
1. - [ ] Add validation check for duplicate endpoint to `odo url create`.
2. Once https://github.com/devfile/api/issues/700 adds changes to devfile to use the endpoint names as container and service port name,
        1. - [ ] add tests to ensure this feature.
        2. - [ ] Add 15 char length validation for url name  to `odo url create` .
